fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneHighlighting.java

```diff
@@ -50,127 +50,214 @@
+                                  TokenIterator standardIt,
+                                  String text,
+                                  boolean cutSnippets,
+                                  int snippetSize,
+                                  @Nullable final List<Pair<Integer, Integer>> highlightedPositions) {
+            this.it = it;
+            this.standardIt = standardIt;
+            this.text = text;
+            this.cutSnippets = cutSnippets;
+            this.snippetSize =  snippetSize;
+            this.highlightedPositions = highlightedPositions;
+        }
+
+        private int tokenCountBeforeHighlighted() {
+            return snippetSize < 0 ? Integer.MAX_VALUE : (snippetSize - 1) / 2;
+        }
+
+        private void handleNonMatchToken() {
+            if (snippetRunningBudget > 0) {
+                addNonMatch(sb, text.substring(upto, standardIt.endOffset()));
+                snippetRunningBudget--;
+            }
+            if (snippetRunningBudget == 0) {
+                beforeHighlightTokens.addLast(text.substring(upto, standardIt.endOffset()));
+                if (beforeHighlightTokens.size() > tokenCountBeforeHighlighted()) {
+                    beforeHighlightTokens.pollFirst();
+                    prefixTextConnector = true;
+                }
+            }
+            upto = standardIt.endOffset();
+        }
+
+        private boolean handleMatchToken(final Set<String> matchedTokens) {
+            if (!matchedTokens.contains(it.getToken())) {
+                return false;
+            }
+            addBeforeTokens();
+            addNonMatch(sb, text.substring(upto, standardIt.startOffset()));
+            addWholeMatch(sb, text.substring(standardIt.startOffset(),standardIt.endOffset()), highlightedPositions);
+            snippetRunningBudget--;
+            upto = it.endOffset();
+            matchedInText.add(standardIt.getToken());
+            return true;
+        }
+
+        private boolean handlePrefixMatch(final Set<String> prefixTokens) {
+            for (String prefixToken : prefixTokens) {
+                if (it.getToken().startsWith(prefixToken)) {
+                    addBeforeTokens();
+                    addNonMatch(sb, text.substring(upto, it.startOffset()));
+                    addPrefixMatch(sb, text.substring(it.startOffset(), it.endOffset()), prefixToken,
+                            highlightedPositions);
+                    upto = it.endOffset();
+                    matchedPrefixes.add(prefixToken);
+                    snippetRunningBudget--;
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private void addBeforeTokens() {
+            snippetRunningBudget = snippetSize < 0 ? beforeHighlightTokens.size()+1 : snippetSize;
+            if (prefixTextConnector) {
+                addNonMatch(sb, highlightedTextConnector);
+                highlightedTextConnector = " ...";
```

Comment:
`" ..."` is I think incorrect for RTL languages, i.e. it should be `"... "` ( relevant 😄 https://www.notarabic.com/ ).